File: python/zap_py/builder.py
"""Builder for ZAP messages — wire-compatible with the Go implementation."""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from typing import List as _PyList, Optional

from .wire import ALIGNMENT, HEADER_SIZE, MAGIC, VERSION

_U16 = struct.Struct("<H")
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")
_I16 = struct.Struct("<h")
_I32 = struct.Struct("<i")
_I64 = struct.Struct("<q")
_F32 = struct.Struct("<f")
_F64 = struct.Struct("<d")


class Builder:
    """Constructs ZAP messages.

    Layout matches builder.go: 16-byte header, then aligned data segment.
    """

    __slots__ = ("_buf", "_pos", "_root_offset")

    def __init__(self, capacity: int = 256):
        if capacity < HEADER_SIZE:
            capacity = 256
        self._buf = bytearray(capacity)
        self._pos = HEADER_SIZE
        self._root_offset = 0
        self._buf[0:4] = MAGIC
        _U16.pack_into(self._buf, 4, VERSION)

    def reset(self) -> None:
        self._pos = HEADER_SIZE
        self._root_offset = 0

    @property
    def pos(self) -> int:
        return self._pos

    def _grow(self, n: int) -> None:
        needed = self._pos + n
        if needed <= len(self._buf):
            return
        new_cap = len(self._buf) * 2
        if new_cap < needed:
            new_cap = needed
        self._buf.extend(bytes(new_cap - len(self._buf)))

    def _align(self, alignment: int) -> None:
        padding = (alignment - (self._pos % alignment)) % alignment
        if padding == 0:
            return
        self._grow(padding)
        for i in range(padding):
            self._buf[self._pos + i] = 0
        self._pos += padding

    def finish(self) -> bytes:
        _U32.pack_into(self._buf, 8, self._root_offset)
        _U32.pack_into(self._buf, 12, self._pos)
        return bytes(self._buf[:self._pos])

    def finish_with_flags(self, flags: int) -> bytes:
        _U16.pack_into(self._buf, 6, flags)
        return self.finish()

    def start_object(self, data_size: int) -> "ObjectBuilder":
        self._align(ALIGNMENT)
        return ObjectBuilder(self, self._pos, data_size)

    def start_list(self, elem_size: int) -> "ListBuilder":
        self._align(ALIGNMENT)
        return ListBuilder(self, self._pos, elem_size)

    def write_bytes(self, data: bytes) -> int:
        if not data:
            return 0
        self._align(ALIGNMENT)
        offset = self._pos
        self._grow(len(data))
        self._buf[self._pos:self._pos + len(data)] = data
        self._pos += len(data)
        return offset

    def write_text(self, s: str) -> int:
        return self.write_bytes(s.encode("utf-8"))


@dataclass
class _OffsetEntry:
    field_offset: int
    data: bytes


class ObjectBuilder:
    """Builds a single ZAP object inside a Builder."""

    __slots__ = ("_b", "_start_pos", "_data_size", "_offsets")

    def __init__(self, b: Builder, start_pos: int, data_size: int):
        self._b = b
        self._start_pos = start_pos
        self._data_size = data_size
        self._offsets: _PyList[_OffsetEntry] = []

    def _ensure_field(self, end_offset: int) -> None:
        needed = self._start_pos + end_offset
        b = self._b
        if needed > b._pos:
            b._grow(needed - b._pos)
            for i in range(b._pos, needed):
                b._buf[i] = 0
            b._pos = needed

    def set_bool(self, field_offset: int, v: bool) -> None:
        self.set_uint8(field_offset, 1 if v else 0)

    def set_uint8(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 1)
        self._b._buf[self._start_pos + field_offset] = v & 0xFF

    def set_uint16(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 2)
        _U16.pack_into(self._b._buf, self._start_pos + field_offset, v & 0xFFFF)

    def set_uint32(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 4)
        _U32.pack_into(self._b._buf, self._start_pos + field_offset, v & 0xFFFFFFFF)

    def set_uint64(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 8)
        _U64.pack_into(self._b._buf, self._start_pos + field_offset, v & 0xFFFFFFFFFFFFFFFF)

    def set_int8(self, field_offset: int, v: int) -> None:
        self.set_uint8(field_offset, v & 0xFF)

    def set_int16(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 2)
        _I16.pack_into(self._b._buf, self._start_pos + field_offset, v)

    def set_int32(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 4)
        _I32.pack_into(self._b._buf, self._start_pos + field_offset, v)

    def set_int64(self, field_offset: int, v: int) -> None:
        self._ensure_field(field_offset + 8)
        _I64.pack_into(self._b._buf, self._start_pos + field_offset, v)

    def set_float32(self, field_offset: int, v: float) -> None:
        self._ensure_field(field_offset + 4)
        _F32.pack_into(self._b._buf, self._start_pos + field_offset, v)

    def set_float64(self, field_offset: int, v: float) -> None:
        self._ensure_field(field_offset + 8)
        _F64.pack_into(self._b._buf, self._start_pos + field_offset, v)

    def set_text(self, field_offset: int, v: str) -> None:
        self.set_bytes(field_offset, v.encode("utf-8"))

    def set_bytes(self, field_offset: int, v: bytes) -> None:
        self._ensure_field(field_offset + 8)
        if len(v) == 0:
            _U32.pack_into(self._b._buf, self._start_pos + field_offset, 0)
            _U32.pack_into(self._b._buf, self._start_pos + field_offset + 4, 0)
            return
        self._offsets.append(_OffsetEntry(field_offset=field_offset, data=bytes(v)))
        _U32.pack_into(self._b._buf, self._start_pos + field_offset + 4, len(v))

    def set_object(self, field_offset: int, obj_offset: int) -> None:
        self._ensure_field(field_offset + 4)
        if obj_offset == 0:
            _U32.pack_into(self._b._buf, self._start_pos + field_offset, 0)
            return
        rel_offset = obj_offset - (self._start_pos + field_offset)
        _I32.pack_into(self._b._buf, self._start_pos + field_offset, rel_offset)

    def set_list(self, field_offset: int, list_offset: int, length: int) -> None:
        self._ensure_field(field_offset + 8)
        if list_offset == 0 or length == 0:
            _U32.pack_into(self._b._buf, self._start_pos + field_offset, 0)
            _U32.pack_into(self._b._buf, self._start_pos + field_offset + 4, 0)
            return
        rel_offset = list_offset - (self._start_pos + field_offset)
        _I32.pack_into(self._b._buf, self._start_pos + field_offset, rel_offset)
        _U32.pack_into(self._b._buf, self._start_pos + field_offset + 4, length)

    def finish(self) -> int:
        self._ensure_field(self._data_size)
        b = self._b
        for entry in self._offsets:
            data_pos = b._pos
            b._grow(len(entry.data))
            b._buf[b._pos:b._pos + len(entry.data)] = entry.data
            b._pos += len(entry.data)
            field_abs_pos = self._start_pos + entry.field_offset
            rel_offset = data_pos - field_abs_pos
            _I32.pack_into(b._buf, field_abs_pos, rel_offset)
        return self._start_pos

    def finish_as_root(self) -> int:
        offset = self.finish()
        self._b._root_offset = offset
        return offset


class ListBuilder:
    """Builds a contiguous list of fixed-width elements."""

    __slots__ = ("_b", "_start_pos", "_elem_size", "_count")

    def __init__(self, b: Builder, start_pos: int, elem_size: int):
        self._b = b
        self._start_pos = start_pos
        self._elem_size = elem_size
        self._count = 0

    def add_uint8(self, v: int) -> None:
        self._b._grow(1)
        self._b._buf[self._b._pos] = v & 0xFF
        self._b._pos += 1
        self._count += 1

    def add_uint32(self, v: int) -> None:
        self._b._grow(4)
        _U32.pack_into(self._b._buf, self._b._pos, v & 0xFFFFFFFF)
        self._b._pos += 4
        self._count += 1

    def add_uint64(self, v: int) -> None:
        self._b._grow(8)
        _U64.pack_into(self._b._buf, self._b._pos, v & 0xFFFFFFFFFFFFFFFF)
        self._b._pos += 8
        self._count += 1

    def add_bytes(self, data: bytes) -> None:
        self._b._grow(len(data))
        self._b._buf[self._b._pos:self._b._pos + len(data)] = data
        self._b._pos += len(data)
        self._count += len(data)

    def finish(self) -> tuple[int, int]:
        return self._start_pos, self._count

File: builder.go
// Copyright (C) 2025, Lux Industries Inc. All rights reserved.
// See the file LICENSE for licensing terms.

package zap

import (
	"encoding/binary"
	"math"
	"sync"
)

// Builder constructs ZAP messages.
type Builder struct {
	buf        []byte
	pos        int
	rootOffset int
}

// NewBuilder creates a new builder with the given initial capacity. The
// resulting message is emitted with Version2 in the wire header (the
// current default). Use NewBuilderV1 only for legacy v1 emitters; new
// code should always use NewBuilder.
func NewBuilder(capacity int) *Builder {
	return newBuilder(capacity, Version2)
}

// NewBuilderV1 creates a new builder that emits Version1 messages. This is
// kept only for explicit-legacy-emitter call sites (none in tree as of
// LP-023 v3.1 round 2). New code should call NewBuilder.
func NewBuilderV1(capacity int) *Builder {
	return newBuilder(capacity, Version1)
}

func newBuilder(capacity int, version uint16) *Builder {
	if capacity < HeaderSize {
		capacity = 256
	}
	b := &Builder{
		buf: make([]byte, capacity),
		pos: HeaderSize, // Start after header
	}
	// Write magic and version
	copy(b.buf[0:4], Magic)
	binary.LittleEndian.PutUint16(b.buf[4:6], version)
	return b
}

// Reset resets the builder for reuse.
func (b *Builder) Reset() {
	b.pos = HeaderSize
	b.rootOffset = 0
}

// grow ensures capacity for n more bytes.
func (b *Builder) grow(n int) {
	if b.pos+n <= len(b.buf) {
		return
	}
	newCap := len(b.buf) * 2
	if newCap < b.pos+n {
		newCap = b.pos + n
	}
	newBuf := make([]byte, newCap)
	copy(newBuf, b.buf[:b.pos])
	b.buf = newBuf
}

// align aligns the current position to the given boundary.
func (b *Builder) align(alignment int) {
	padding := (alignment - (b.pos % alignment)) % alignment
	b.grow(padding)
	for i := 0; i < padding; i++ {
		b.buf[b.pos] = 0
		b.pos++
	}
}

// Finish finalizes the message and returns the bytes.
func (b *Builder) Finish() []byte {
	// Write root offset
	binary.LittleEndian.PutUint32(b.buf[8:12], uint32(b.rootOffset))
	// Write total size
	binary.LittleEndian.PutUint32(b.buf[12:16], uint32(b.pos))
	return b.buf[:b.pos]
}

// FinishWithFlags finalizes with specific flags.
func (b *Builder) FinishWithFlags(flags uint16) []byte {
	binary.LittleEndian.PutUint16(b.buf[6:8], flags)
	return b.Finish()
}

// ObjectBuilder builds a ZAP object (struct).
type ObjectBuilder struct {
	b        *Builder
	startPos int
	dataSize int
}

// StartObject starts building an object with the given data size, RESERVING the
// full fixed section up front (zero-filled). Reserving eagerly means a variable
// field (SetBytes/SetText) can append its tail data immediately after the fixed
// section and patch its own relative pointer on the spot — there is no deferred
// offset list to record intentions and replay in Finish. Callers already build
// child objects/lists BEFORE StartObject (the writeXxxList-then-StartObject
// pattern), so the object's own Set* calls are the only buffer writes between
// StartObject and Finish; appending tail bytes immediately is safe and produces
// byte-identical wire.
func (b *Builder) StartObject(dataSize int) ObjectBuilder {
	b.align(Alignment)
	ob := ObjectBuilder{
		b:        b,
		startPos: b.pos,
		dataSize: dataSize,
	}
	ob.ensureField(dataSize)
	return ob
}

// SetBool sets a bool field.
func (ob ObjectBuilder) SetBool(fieldOffset int, v bool) {
	if v {
		ob.SetUint8(fieldOffset, 1)
	} else {
		ob.SetUint8(fieldOffset, 0)
	}
}

// SetUint8 sets a uint8 field.
func (ob ObjectBuilder) SetUint8(fieldOffset int, v uint8) {
	ob.ensureField(fieldOffset + 1)
	ob.b.buf[ob.startPos+fieldOffset] = v
}

// SetUint16 sets a uint16 field.
func (ob ObjectBuilder) SetUint16(fieldOffset int, v uint16) {
	ob.ensureField(fieldOffset + 2)
	binary.LittleEndian.PutUint16(ob.b.buf[ob.startPos+fieldOffset:], v)
}

// SetUint32 sets a uint32 field.
func (ob ObjectBuilder) SetUint32(fieldOffset int, v uint32) {
	ob.ensureField(fieldOffset + 4)
	binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], v)
}

// SetUint64 sets a uint64 field.
func (ob ObjectBuilder) SetUint64(fieldOffset int, v uint64) {
	ob.ensureField(fieldOffset + 8)
	binary.LittleEndian.PutUint64(ob.b.buf[ob.startPos+fieldOffset:], v)
}

// SetInt8 sets an int8 field.
func (ob ObjectBuilder) SetInt8(fieldOffset int, v int8) {
	ob.SetUint8(fieldOffset, uint8(v))
}

// SetInt16 sets an int16 field.
func (ob ObjectBuilder) SetInt16(fieldOffset int, v int16) {
	ob.SetUint16(fieldOffset, uint16(v))
}

// SetInt32 sets an int32 field.
func (ob ObjectBuilder) SetInt32(fieldOffset int, v int32) {
	ob.SetUint32(fieldOffset, uint32(v))
}

// SetInt64 sets an int64 field.
func (ob ObjectBuilder) SetInt64(fieldOffset int, v int64) {
	ob.SetUint64(fieldOffset, uint64(v))
}

// SetFloat32 sets a float32 field.
func (ob ObjectBuilder) SetFloat32(fieldOffset int, v float32) {
	ob.SetUint32(fieldOffset, float32bits(v))
}

// SetFloat64 sets a float64 field.
func (ob ObjectBuilder) SetFloat64(fieldOffset int, v float64) {
	ob.SetUint64(fieldOffset, float64bits(v))
}

// SetBytesFixed copies len(v) bytes inline at fieldOffset within the
// object's fixed payload. This is the generic-width counterpart of
// SetHash (32 bytes) and SetAddress (20 bytes); it covers any N>0
// inline byte-array slot (e.g., LP-201 SessionID [16]byte, LP-208
// QuasarWitness [96]byte).
//
// Unlike SetBytes (which writes a variable-length tail pointer
// {relOffset uint32, length uint32}), SetBytesFixed writes the bytes
// IN PLACE in the fixed payload. Use this when the schema declares a
// fixed-width byte field; use SetBytes when the schema declares a
// variable-length tail.
//
// Panics on len(v) == 0 are avoided: a zero-length argument is a
// no-op (the slot is left as the zero value).
func (ob ObjectBuilder) SetBytesFixed(fieldOffset int, v []byte) {
	if len(v) == 0 {
		return
	}
	ob.ensureField(fieldOffset + len(v))
	copy(ob.b.buf[ob.startPos+fieldOffset:], v)
}

// SetText sets a text (string) field.
func (ob ObjectBuilder) SetText(fieldOffset int, v string) {
	ob.SetBytes(fieldOffset, []byte(v))
}

// SetBytes sets a bytes field.
// The data is written after the object's fixed section during Finish().
func (ob ObjectBuilder) SetBytes(fieldOffset int, v []byte) {
	if len(v) == 0 {
		// Null pointer (offset 0, length 0).
		binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], 0)
		binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset+4:], 0)
		return
	}
	// The fixed section is fully reserved by StartObject, so b.pos is at the
	// tail: append the data now and patch this field's (relOffset, length) pair
	// immediately — no deferred offset list, no per-object slice allocation.
	dataPos := ob.b.pos
	ob.b.grow(len(v))
	copy(ob.b.buf[ob.b.pos:ob.b.pos+len(v)], v)
	ob.b.pos += len(v)

	fieldAbsPos := ob.startPos + fieldOffset
	relOffset := int32(dataPos - fieldAbsPos)
	binary.LittleEndian.PutUint32(ob.b.buf[fieldAbsPos:], uint32(relOffset))
	binary.LittleEndian.PutUint32(ob.b.buf[fieldAbsPos+4:], uint32(len(v)))
}

// SetObject sets a nested object field (by offset).
func (ob ObjectBuilder) SetObject(fieldOffset int, objOffset int) {
	ob.ensureField(fieldOffset + 4)

	if objOffset == 0 {
		binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], 0)
		return
	}

	// Calculate relative offset from field position
	relOffset := objOffset - (ob.startPos + fieldOffset)
	binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], uint32(relOffset))
}

// SetList sets a list field.
func (ob ObjectBuilder) SetList(fieldOffset int, listOffset int, length int) {
	ob.ensureField(fieldOffset + 8)

	if listOffset == 0 || length == 0 {
		binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], 0)
		binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset+4:], 0)
		return
	}

	relOffset := listOffset - (ob.startPos + fieldOffset)
	binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset:], uint32(relOffset))
	binary.LittleEndian.PutUint32(ob.b.buf[ob.startPos+fieldOffset+4:], uint32(length))
}

func (ob ObjectBuilder) ensureField(endOffset int) {
	needed := ob.startPos + endOffset
	if needed > ob.b.pos {
		ob.b.grow(needed - ob.b.pos)
		// Zero-fill
		for i := ob.b.pos; i < needed; i++ {
			ob.b.buf[i] = 0
		}
		ob.b.pos = needed
	}
}

// ReserveFixed extends the builder's write cursor so the object's
// entire fixed payload of dataSize bytes is materialized (zero-filled
// up to startPos+dataSize). Use this BEFORE writing any variable-
// length tail (list elements, deferred bytes) that should live AFTER
// the fixed section.
//
// Without this call, list elements or deferred-data writes interleave
// with the unreserved tail of the fixed section, producing incorrect
// wire bytes when later Set* calls patch fields that overlap with
// already-written variable data.
//
// Idempotent: a second call with the same dataSize is a no-op. A call
// with a smaller dataSize is also a no-op (the cursor never moves
// backwards).
//
// This is the exported counterpart to the internal ensureField helper.
// It is used by zapv1.WriteList to keep the parent's payload reserved
// before list elements are appended.
func (ob ObjectBuilder) ReserveFixed(dataSize int) {
	ob.ensureField(dataSize)
}

// Finish finalizes the object and returns its offset. Every field (fixed and
// variable) is written eagerly by its Set* call — the fixed section is reserved
// in StartObject and SetBytes/SetText append their tail immediately — so there
// is nothing to replay here.
func (ob ObjectBuilder) Finish() int {
	return ob.startPos
}

// FinishAsRoot finalizes and sets as the message root.
func (ob ObjectBuilder) FinishAsRoot() int {
	offset := ob.Finish()
	ob.b.rootOffset = offset
	return offset
}

// WriteBytes writes raw bytes and returns the offset.
func (b *Builder) WriteBytes(data []byte) int {
	if len(data) == 0 {
		return 0
	}
	b.align(Alignment)
	offset := b.pos
	b.grow(len(data))
	copy(b.buf[b.pos:], data)
	b.pos += len(data)
	return offset
}

// WriteText writes a string and returns the offset.
func (b *Builder) WriteText(s string) int {
	return b.WriteBytes([]byte(s))
}

// ListBuilder builds a ZAP list.
type ListBuilder struct {
	b        *Builder
	startPos int
	count    int
}

// StartList starts building a list. Returns a VALUE (not a heap pointer): the
// Add*/Finish methods take pointer receivers, and every caller binds the result
// to an addressable local (`lb := b.StartList(...)`), so Go auto-addresses the
// local and count mutations accumulate on it — no per-list heap allocation.
// Mirrors the value-typed StartObject. elemSize is unused (the stride is
// implicit in which Add* the caller invokes) and kept only as a documenting
// param.
func (b *Builder) StartList(elemSize int) ListBuilder {
	_ = elemSize
	b.align(Alignment)
	return ListBuilder{
		b:        b,
		startPos: b.pos,
	}
}

// AddUint8 adds a uint8 element.
func (lb *ListBuilder) AddUint8(v uint8) {
	lb.b.grow(1)
	lb.b.buf[lb.b.pos] = v
	lb.b.pos++
	lb.count++
}

// AddUint32 adds a uint32 element.
func (lb *ListBuilder) AddUint32(v uint32) {
	lb.b.grow(4)
	binary.LittleEndian.PutUint32(lb.b.buf[lb.b.pos:], v)
	lb.b.pos += 4
	lb.count++
}

// AddUint64 adds a uint64 element.
func (lb *ListBuilder) AddUint64(v uint64) {
	lb.b.grow(8)
	binary.LittleEndian.PutUint64(lb.b.buf[lb.b.pos:], v)
	lb.b.pos += 8
	lb.count++
}

// AddBytes adds raw bytes (for byte lists).
func (lb *ListBuilder) AddBytes(data []byte) {
	lb.b.grow(len(data))
	copy(lb.b.buf[lb.b.pos:], data)
	lb.b.pos += len(data)
	lb.count += len(data)
}

// AddObjectPtr appends a 4-byte SIGNED relative pointer to an object at
// absolute position targetPos (pass 0 for a null element). This is the
// element kind for a list of out-of-line objects — the wire encoding of a
// "repeated message" field, where each fixed-stride slot points to a tailed
// object elsewhere in the buffer rather than inlining a fixed-size value.
//
// The relative offset is computed against THIS slot's position, so the
// reader resolves it the same way [Object.Object] does: absOffset = slotPos
// + int32(rel). Targets may precede the slot (negative rel) — the common
// case, since a builder writes the objects first and the pointer array
// after.
func (lb *ListBuilder) AddObjectPtr(targetPos int) {
	lb.b.grow(4)
	if targetPos == 0 {
		binary.LittleEndian.PutUint32(lb.b.buf[lb.b.pos:], 0)
	} else {
		rel := int32(targetPos - lb.b.pos)
		binary.LittleEndian.PutUint32(lb.b.buf[lb.b.pos:], uint32(rel))
	}
	lb.b.pos += 4
	lb.count++
}

// Finish returns the list offset and length.
func (lb *ListBuilder) Finish() (offset int, length int) {
	return lb.startPos, lb.count
}

// ---- Builder pool (write-side counterpart to the read bufpool) ----

// builderPool recycles Builders across messages. A pooled Builder keeps its
// grown backing array, so steady-state message construction allocates zero
// builder buffers. Byte-safety on reuse is unconditional: ensureField
// zero-fills every extended span and align zero-fills padding, so a reused
// (dirty) buffer emits bytes identical to a fresh one.
var builderPool = sync.Pool{
	New: func() any { return NewBuilder(512) },
}

// GetBuilder returns a pooled Builder, reset and ready to write a Version2
// message. Callers MUST copy out the bytes returned by Finish (which aliases
// the Builder's buffer) before calling PutBuilder.
func GetBuilder() *Builder {
	b := builderPool.Get().(*Builder)
	b.Reset()
	return b
}

// PutBuilder returns a Builder to the pool. The slice previously returned by
// b.Finish() must no longer be referenced (it aliases b's buffer).
func PutBuilder(b *Builder) {
	builderPool.Put(b)
}

// Helper functions

func float32bits(f float32) uint32 {
	return math.Float32bits(f)
}

func float64bits(f float64) uint64 {
	return math.Float64bits(f)
}
